src/GafferScene/IECoreScenePreview/PrimitiveAlgo.cpp

```diff
@@ -0,0 +1,1188 @@
+		}
+
+		result = resultMesh;
+	}
+	else if( resultTypeId == IECoreScene::CurvesPrimitiveTypeId )
+	{
+		resultCurves = new IECoreScene::CurvesPrimitive();
+		result = resultCurves;
+
+		resultVerticesPerCurveData = new IntVectorData;
+		resultVerticesPerCurveData->writable().resize( totalInterpolation[ PrimitiveVariable::Uniform ] );
+	}
+	else
+	{
+		result = new IECoreScene::PointsPrimitive( totalInterpolation[ PrimitiveVariable::Vertex ] );
+	}
+
+	//
+	// Allocate storage for the primitives variables
+	//
+
+	for( auto &[name, varInfo] : varInfos )
+	{
+		if( varInfo.invalid )
+		{
+			continue;
+		}
+
+		varInfo.accumDataSizes.reserve( varInfo.numData.size() );
+		size_t accumDataSize = 0;
+		for( unsigned int i : varInfo.numData )
+		{
+			varInfo.accumDataSizes.push_back( accumDataSize );
+			accumDataSize += i;
+		}
+
+		PrimitiveVariable &p = result->variables.emplace(name, PrimitiveVariable() ).first->second;
+
+		p.data = IECore::runTimeCast<Data>( IECore::Object::create( varInfo.typeId ) );
+
+		IECore::setGeometricInterpretation( p.data.get(), varInfo.interpret );
+		p.interpolation = varInfo.interpolation;
+		Canceller::check( canceller );
+		dataResize( p.data.get(), accumDataSize );
+
+		if( varInfo.indexed )
+		{
+			p.indices = new IntVectorData();
+			Canceller::check( canceller );
+			p.indices->writable().resize( totalInterpolation[ varInfo.interpolation ] );
+		}
+	}
+
+	//
+	// Now a big parallel loop where we do the majority of actual work - copying all the primvar and topology
+	// data to the destination.
+	//
+
+	tbb::task_group_context taskGroupContext( tbb::task_group_context::isolated );
+
+	tbb::parallel_for(
+		tbb::blocked_range<size_t>( 0, primitives.size() ),
+		[&]( tbb::blocked_range<size_t> &range )
+		{
+			for( size_t i = range.begin(); i != range.end(); i++ )
+			{
+				const Primitive &sourcePrim = *primitives[i].first;
+
+				const Imath::M44f &matrix = primitives[i].second;
+				const Imath::M44f normalMatrix = normalTransform( matrix );
+
+				// Copy the data ( and indices ) for each prim var for this primitive into
+				// the destination primvar.
+				for( auto &[name, varInfo] : varInfos )
+				{
+					if( varInfo.invalid )
+					{
+						continue;
+					}
+
+					PrimitiveVariable &destVar = result->variables.find( name )->second;
+
+					const size_t numIndices = countInterpolation[ varInfo.interpolation ][i];
+					const size_t startIndex = accumInterpolation[ varInfo.interpolation ][i];
+					const size_t dataStart = varInfo.accumDataSizes[i];
+
+
+					auto it = sourcePrim.variables.find( name );
+					if( it == sourcePrim.variables.end() || it->second.interpolation == PrimitiveVariable::Invalid )
+					{
+						// No matching data found in this primitive for this primvar
+
+						// We don't currently have a way to suppress zero-initialization of the data, so
+						// we don't need to initialize that here
+
+						if( varInfo.indexed )
```

Comment:
If I understand the earlier phase correctly, I _think_ `varInfo.indexed` should be guaranteed to be `true` here. Is that not the case?

Reply:
Good point, addressed in 966690e2d0d3660edd266c6dff7c01baa81a4fb6